cpp/include/cudf/utilities/pinned_allocator.hpp

```diff
@@ -0,0 +1,223 @@
+  using pointer         = T*;              ///< The type returned by address() / allocate()
+  using const_pointer   = const T*;        ///< The type returned by address()
+  using reference       = T&;              ///< The parameter type for address()
+  using const_reference = const T&;        ///< The parameter type for address()
+  using size_type       = std::size_t;     ///< The type used for the size of the allocation
+  using difference_type = std::ptrdiff_t;  ///< The type of the distance between two pointers
+
+  /**
+   * @brief converts a `pinned_allocator<T>` to `pinned_allocator<U>`
+   */
+  template <typename U>
+  struct rebind {
+    using other = pinned_allocator<U>;  ///< The rebound type
+  };
+
+  /**
+   * @brief pinned_allocator's null constructor does nothing.
+   */
+  __host__ __device__ inline pinned_allocator() {}
+
+  /**
+   * @brief pinned_allocator's null destructor does nothing.
+   */
+  __host__ __device__ inline ~pinned_allocator() {}
+
+  /**
+   * @brief pinned_allocator's copy constructor does nothing.
+   */
+  __host__ __device__ inline pinned_allocator(pinned_allocator const&) {}
+
+  /**
+   * @brief  pinned_allocator's copy constructor does nothing.
+   *
+   *  This version of pinned_allocator's copy constructor
+   *  is templated on the \c value_type of the pinned_allocator
+   *  to copy from.  It is provided merely for convenience; it
+   *  does nothing.
+   */
+  template <typename U>
+  __host__ __device__ inline pinned_allocator(pinned_allocator<U> const&)
+  {
+  }
+
+  /**
+   * @brief This method returns the address of a \c reference of
+   *  interest.
+   *
+   *  @param r The \c reference of interest.
+   *  @return \c r's address.
+   */
+  __host__ __device__ inline pointer address(reference r) { return &r; }
+
+  /**
+   * @brief This method returns the address of a \c const_reference
+   *  of interest.
+   *
+   *  @param r The \c const_reference of interest.
+   *  @return \c r's address.
+   */
+  __host__ __device__ inline const_pointer address(const_reference r) { return &r; }
+
+  /**
+   * @brief This method allocates storage for objects in pinned host
+   *  memory.
+   *
+   *  @param cnt The number of objects to allocate.
+   *  @return a \c pointer to the newly allocated objects.
+   *  @note The second parameter to this function is meant as a
+   *        hint pointer to a nearby memory location, but is
+   *        not used by this allocator.
+   *  @note This method does not invoke \p value_type's constructor.
+   *        It is the responsibility of the caller to initialize the
+   *        objects at the returned \c pointer.
+   */
+  __host__ inline pointer allocate(size_type cnt, const_pointer /*hint*/ = 0)
+  {
+    if (cnt > this->max_size()) { throw std::bad_alloc(); }  // end if
+
+    pointer result(0);
+    cudaError_t error = cudaMallocHost(reinterpret_cast<void**>(&result), cnt * sizeof(value_type));
```

Comment:
I know this is just copy/paste from Thrust, but we should probably do a bit more than that and use things like `CUDF_CUDA_TRY` for error checking and throw `cudf::` specific exceptions.

Reply:
I think that would eliminate the need for this header to include `thrust` headers as well, especially `detail` headers.

Reply:
Would throwing `cudf::` exceptions have any impact on this class being used by `thrust::host_vector` does it expect that all execptions throwin are derived from `thrust` exceptions?

Reply:
> expect that all execptions throwin are derived from thrust exceptions

I would certainly hope not. The `Allocator` is intended to user provided and there should be no exception to derive from anything other than the appropriate `std::` exceptions.